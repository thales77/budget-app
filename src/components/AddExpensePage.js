import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push('/dashboard');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            categories={this.props.categories}
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
    categories: state.categories.map(({ description }) => description)
});

//we pass dispatch through props
//theoretically we already have access to dispatch
//but by using mapDispatchToProps we decouple the component from 
//Redux, and thus making the component more reusable 
const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);
