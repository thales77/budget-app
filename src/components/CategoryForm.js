import React from 'react';


export default class CategoryForm extends React.Component {
    constructor(props) {
        super(props);

        //if we are editing, setup state from props passed by parent component
        //else if we are adding a new category setup empty state
        this.state = {
            description: props.category ? props.category.description : '',
            note: props.category ? props.category.note : '',
            budgetAmount: props.category ? (props.category.budgetAmount / 100).toString() : '',
            error: ''
        };
    };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onBudgetAmountChange = (e) => {
        const budgetAmount = e.target.value;

        if (!budgetAmount || budgetAmount.match(/^\d{1,}(\,\d{0,2})?$/)) {
            this.setState(() => ({ budgetAmount }));
        }
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.budgetAmount) {
            this.setState(() => ({ error: 'Please provide description and amount' }))
        } else {
            this.setState(() => ({ error: '' }));

            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                //parsefloat only accepts . for decimal 
                budgetAmount: parseFloat((this.state.budgetAmount).replace(',', '.'), 10) * 100,
            });
        }
    }
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    className="text-input"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    placeholder="Budget amount"
                    className="text-input"
                    value={this.state.budgetAmount}
                    onChange={this.onBudgetAmountChange}
                />
                <textarea
                    className="textarea"
                    placeholder="Add a note for your category (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                <div>
                    <button className="button">Save Category</button>
                </div>
            </form>
        );
    };
};