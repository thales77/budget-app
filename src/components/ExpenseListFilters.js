import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

// AirBnB Date picker imports
import { DateRangePicker } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css';
// import 'react-dates/initialize';


export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
    };
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            placeholder="Search expenses"
                            className="text-input"
                            type="text"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            startDateId={'a'}
                            endDateId={'b'}
                        />
                    </div>
                </div>
            </div>
        );
    }
}


const mapStateToProps = state => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: startDate => dispatch(setStartDate(startDate)),
    setEndDate: endDate => dispatch(setEndDate(endDate)),
    setTextFilter: text => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);