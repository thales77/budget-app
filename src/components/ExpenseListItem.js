import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt, note }) => (
    <Link className="list-item" to={`/editExpense/${id}`}>
        <div>
            <h3 className="list-item__title">{description}</h3>
            <span className="list-item__subtitle">{moment(createdAt).format('DD MMM YYYY')}</span>
        </div>
        <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
    </Link>
);

export default ExpenseListItem;