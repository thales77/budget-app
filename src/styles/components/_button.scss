
.button {
    background-color: $blue;
    border: none;
    color: white;
    display: inline-block;
    font-weight: 300;
    font-size: $font-size-large;
    line-height: 1;
    padding: $s-size;
    text-decoration: none;
    margin: 0 0.5rem 0 0;
}

.button--link {
    background: none;
}

.button--secondary {
    background: #888888;
}