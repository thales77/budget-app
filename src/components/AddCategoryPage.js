import React from 'react';
import { connect } from 'react-redux';
import CategoryForm from './CategoryForm';
import { startAddCategory } from '../actions/categories';

export class AddCategoryPage extends React.Component {
  onSubmit = category => {
    this.props.startAddCategory(category);
    this.props.history.push('/dashboard');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Category</h1>
          </div>
        </div>
        <div className="content-container">
          <CategoryForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

//we pass dispatch through props
//theoretically we already have access to dispatch
//but by using mapDispatchToProps we decouple the component from 
//Redux, and thus making the component more reusable 
const mapDispatchToProps = dispatch => ({
  startAddCategory: category => dispatch(startAddCategory(category))
});

export default connect(undefined, mapDispatchToProps)(AddCategoryPage);
