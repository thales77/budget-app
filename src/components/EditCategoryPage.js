import React from 'react';
import { connect } from 'react-redux';
import CategoryForm from './CategoryForm';
import { startEditCategory, startRemoveCategory } from '../actions/categories';

export class EditCategoryPage extends React.Component {
  onSubmit = category => {
    //pass this to store
    this.props.startEditCategory(this.props.category.id, category);
    this.props.history.push('/dashboard');
  };
  onRemove = () => {
    this.props.startRemoveCategory({ id: this.props.category.id });
    this.props.history.push('/dashboard');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit category</h1>
          </div>
        </div>
        <div className="content-container">
          <CategoryForm category={this.props.category} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.onRemove}>Remove category </button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  category: state.categories.find(category => category.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditCategory: (id, category) => dispatch(startEditCategory(id, category)),
  startRemoveCategory: data => dispatch(startRemoveCategory(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditCategoryPage);