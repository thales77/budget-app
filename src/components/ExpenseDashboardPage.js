import React from 'react';
import ExpenseSummary from './ExpensesSummary';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseList from './ExpenseList';
import setCurrencyLocale from '../utils/setCurrencyLocale';

setCurrencyLocale('it');

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;