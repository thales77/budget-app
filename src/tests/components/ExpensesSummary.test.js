import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import getExpensesTotal from '../../selectors/expenses-total';
import expenses from '../fixtures/expenses';

const expenseCount = expenses.length;
const expensesTotal = getExpensesTotal(expenses);

test('Render expense summary item', () => {
    const wrapper = shallow(
        < ExpensesSummary
            expenseCount={expenseCount}
            expensesTotal={expensesTotal}
        />
    );
    expect(wrapper).toMatchSnapshot();
});