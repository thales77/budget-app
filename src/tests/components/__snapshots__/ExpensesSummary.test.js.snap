// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render expense summary item 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
         
      </span>
      expenses
       totalling 
      <span>
        $1,141.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <[object Object]
        className="button"
        to="/create"
      >
        Add Expense
      </[object Object]>
    </div>
  </div>
</div>
`;
